src/orchestration/training_cluster_monitor.py

```diff
@@ -30,93 +30,128 @@
+  )
 
-    os.environ['ACTION'] = 'CREATE'
+  os.environ['ACTION'] = 'CREATE'
 
-    logging.info("Starting cluster: %s", os.environ['NAME_PREFIX'])
-    subprocess.run(["bash", "/usr/entrypoint.sh"],
-        capture_output=True, check=False)
+  logging.info('Starting cluster: %s', os.environ['NAME_PREFIX'])
+  subprocess.run(
+      ['bash', '/usr/entrypoint.sh'], capture_output=True, check=False
+  )
 
 
 def main(argv):
-    if len(argv) > 1:
-        raise app.UsageError('Too many command-line arguments.')
-
-    logging.info("Project id: %s", FLAGS.project_id)
-    logging.info("train_ttl_sec: %s", FLAGS.train_ttl_sec)
-    logging.info("Checkpoint Path: %s", FLAGS.checkpoint_path)
-
-    os.environ["GCLOUD_PROJECT"] = FLAGS.project_id
-
-    bucket_name, checkpoint_prefix = _parse_gcs_resource_name(
-        FLAGS.checkpoint_path)
-
-    while True:
-        if not _is_process_running(parent_id):
-            sys.exit(0)
-        latest_checkpoint_time = _get_latest_file_utime(
-            bucket_name, checkpoint_prefix)
-        if latest_checkpoint_time is None:
-            latest_checkpoint_time = _get_latest_file_utime(
-                bucket_name,os.path.dirname(checkpoint_prefix))
-        logging.info("Latest checkpoint file time: %s", latest_checkpoint_time)
-
-        if (latest_checkpoint_time is not None and
-            _should_restart_training_cluster(
-                latest_checkpoint_time, FLAGS.train_ttl_sec)):
-            logging.info(
-                "Restarts training cluster, latest checkpoint too old at: %s",
-                latest_checkpoint_time)
-            _restart_training_cluster()
-        time.sleep(FLAGS.polling_sec)
+  if len(argv) > 1:
+    raise app.UsageError('Too many command-line arguments.')
+
+  logging.info('Project id: %s', FLAGS.project_id)
+  logging.info('train_ttl_sec: %s', FLAGS.train_ttl_sec)
+  logging.info('Checkpoint Path: %s', FLAGS.checkpoint_path)
+
+  os.environ['GCLOUD_PROJECT'] = FLAGS.project_id
+
+  bucket_name, model_output_prefix = _parse_gcs_resource_name(
+      FLAGS.model_output
+  )
+
+  num_reloads = 0
+
+  while True:
+    if not _is_process_running(parent_id):
+      sys.exit(0)
+    latest_checkpoint_time = _get_latest_checkpoint_file_utime(
+        bucket_name, model_output_prefix
+    )
+
+    if latest_checkpoint_time is None:
+      latest_checkpoint_time = _get_latest_file_utime(
+          bucket_name, model_output_prefix
+      )
+    logging.info('Latest checkpoint file time: %s', latest_checkpoint_time)
+
+    if latest_checkpoint_time is not None and _should_restart_training_cluster(
+        latest_checkpoint_time,
+        FLAGS.train_ttl_sec,
+        num_reloads,
+        FLAGS.max_reloads,
+    ):
+      logging.info(
+          'Restarts training cluster, latest checkpoint too old at: %s',
+          latest_checkpoint_time,
+      )
+      _restart_training_cluster()
+      num_reloads += 1
```

Comment:
It would be reasonable to terminated the pipeline when we reach maximum reloads. Termination can be as easy as echoing 'failed' into progress.txt

Reply:
Done